File: main2.py
import json
import click

def printScreen(screen, Ht, Lt):
    for y in range(Ht):
        for x in range(Lt):
            print(screen[y][x], end="")
        print()
def drawBox2(screen, Ht, Lt, d, offset, last):
    l=d[0]
    w=d[1]
    h=d[2]

    # Front face
    ym = Ht-h-2
    xm = offset+l+2
    yt = ym-w-1
    xt = xm+w+1
    on = l+w+4+offset

    for y in range(yt, Ht):
        for x in range(offset, on):
            if y >= ym and x < xm:
                if y == ym:
                    c = '+' if x == offset or x == xm - 1 else '-'
                elif y == Ht - 1:
                    c = '+' if x == offset or x == xm - 1 else '-'
                else:
                    c = '|' if x == offset or x == xm - 1 else ' '
                screen[y][x] = c
            if y >= yt and y < ym and x < xt:
                xs = offset + w + 1 - (y - yt)
                xf = xs + l + 1
                if y == yt:
                    if x >= xs:
                        c = '+' if x == xs or x == xf else '-'
                    else:
                        c = ' '
                else:
                    c = '/' if x == xs or x == xf else ' '
                screen[y][x] = c
            if y >= yt+1 and y < yt+1+h and x == xt-1:
                c = '|'
                screen[y][x] = c
            if y >= yt+h+1 and y < Ht:
                xs = xm+w-y+yt+h+1
                xf = xs+y-yt-h
                xlimit = min(xf+1, Lt)
                if x >= xs and x < xlimit:
                    if y == yt+h+1:
                        c = '+' if x == xs else ('_' if not last else ' ')
                    else:
                        if y == Ht-1 and x==xs:
                            continue
                        c = '/' if x == xs else ('_' if not last else ' ')
                    screen[y][x] = c

    return on
@click.command()
@click.option('--dims')
def main(dims):
    ds = json.loads(dims)
    w = ds[0][1]
    n = len(ds)
    Lt = sum([d[0] for d in ds]) + n*w + n*4 - 1
    Ht = max([d[2] for d in ds]) + 2 + w + 1

    screen = [[' ' for x in range(Lt)] for y in range(Ht)]
    offset = 0
    for i,d in enumerate(ds):
        offset = drawBox2(screen, Ht, Lt, d, offset, i == n-1)
    printScreen(screen, Ht, Lt)

if __name__ == '__main__':
    main()

File: main2.rb
require 'json'
def I(p,q)x=p;while G[q][x]==" ";G[q][x]="_";x-=1;end;end
def C(p,q,h,f,b=nil)a=q+h+1;(q..a).each{|y|G[f==1 ?y:p][f==1 ? p:y]=y==q||y==a ? "+":f==1 ? "|":"-";if y>=U&&b;I(p-1,y)end}end
def D(p,q,b=nil)a=q+W+1;(q..a).each{|y|G[y][p-y+q]=y==q||y==a ? "+":"/";if y>=U&&b;I(p-y+q-1,y)end}end
def E(d,o)l=d[0];h=d[2];a=H-h-2;b=o+l+1;c=a-W-1;e=b+W+1;f=o+W;r=l+f+4;C(H-1,o,l,0);C(a,o,l,0);C(c,f+1,l,0);C(o,a,h,1,o>0);C(b,a,h,1);C(e,c,h,1);D(f+1,c,o>0);D(e,c);D(e,U);r;end
f=JSON.parse(ARGV[0])
W=f[0][1]
n=f.size
H=f.map{|d|d[2]}.max+3+W
U=H-2-W
G=Array.new(H){Array.new(f.map{|d|d[0]}.sum+n*W+n*4-1){" "}};o=0
f.each{|d|o=E d,o}
puts G.map{|y|y.join("")}
